source/lac/sparsity_tools.cc

```diff
@@ -39,15 +39,258 @@ extern "C"
+    //Query functions for partition_zoltan
+    int get_number_of_objects(void *data, int *ierr)
+    {
+      SparsityPattern *graph = static_cast<SparsityPattern *>(data);
+      *ierr = ZOLTAN_OK;
+
+      return graph->n_rows();
+    }
+
+
+    void get_object_list(void *data, int sizeGID, int sizeLID,
+                         ZOLTAN_ID_PTR globalID, ZOLTAN_ID_PTR localID,
+                         int wgt_dim, float *obj_wgts, int *ierr)
+    {
+      SparsityPattern *graph = static_cast<SparsityPattern *>(data);
+      *ierr = ZOLTAN_OK;
+
+      Assert( globalID != nullptr , ExcEmptyObject() );
+      Assert( localID != nullptr , ExcEmptyObject() );
+
+      //set global degrees of freedom
+      auto n_dofs = graph->n_rows();
+      for (int i=0; i < n_dofs; i++)
+        {
+          globalID[i] = i;    //global ids run from 0 to n_dofs-1
+          localID[i] = i;     //same as global ids
+        }
+      return;
+    }
+
+
+    void get_num_edges_list(void *data, int sizeGID, int sizeLID,
+                            int num_obj,
+                            ZOLTAN_ID_PTR globalID, ZOLTAN_ID_PTR localID,
+                            int *numEdges, int *ierr)
+    {
+      SparsityPattern *graph = static_cast<SparsityPattern *>(data);
+      *ierr = ZOLTAN_OK;
+
+      Assert ( numEdges != nullptr , ExcEmptyObject() );
+
+      for (int i=0; i<num_obj; ++i)
+        numEdges[i] = graph->row_length(globalID[i])-1;
+    }
+
+
+    void get_edge_list(void *data, int sizeGID, int sizeLID,
+                       int num_obj, ZOLTAN_ID_PTR globalID, ZOLTAN_ID_PTR localID,
+                       int *num_edges,
+                       ZOLTAN_ID_PTR nborGID, int *nborProc,
+                       int wgt_dim, float *ewgts, int *ierr)
+    {
+      SparsityPattern *graph = static_cast<SparsityPattern *>(data);
+      *ierr = ZOLTAN_OK;
+
+      ZOLTAN_ID_PTR nextNborGID = nborGID;
+      int *nextNborProc = nborProc;
+
+      Assert( nextNborGID != nullptr , ExcEmptyObject() );
+      Assert( nextNborProc != nullptr , ExcEmptyObject() );
+
+      //Loop through rows corresponding to indices in globalID implicitly
+      for (int i=0; i < num_obj; ++i)
+        {
+          //Loop through each column to find neighbours
+          for ( SparsityPattern::iterator col=graph->begin(i);
+                col < graph->end(i); ++col )
+            //Ignore diagonal entries. Not needed for partitioning.
+            if ( i != col->column() )
+              {
+                *nextNborGID++ = col->column();
+                *nextNborProc++ = 0;    //All the vertices on processor 0
+              }
+        }
+    }
+
+
+    void partition_zoltan (const SparsityPattern     &sparsity_pattern,
+                           const unsigned int         n_partitions,
+                           std::vector<unsigned int> &partition_indices)
+    {
+      //MPI environment must have been initialized by this point.
+      Zoltan *zz = new Zoltan(MPI::COMM_SELF);
```

Comment:
use
```
  std::unique_ptr<Zoltan> zz = std_cxx14::make_unique<Zoltan>(MPI::COMM_SELF));
```
here. You then don't need the `delete` below, and you also don't need the check for `nullptr` in the next line -- `new` (or `make_unique`) will throw an exception if it can't allocate the memory.

Reply:
Further, this should probably be `MPI_COMM_SELF`.

Reply:
Oh, yes, good eyes! The MPI C++ bindings are going to go away.